### PR TITLE
Gh 1252 - Provide a means for ANET administrators to return engagement reports to report authors

### PR DESCRIPTION
**Pull request by maradragan**
- From now on ANET administrators can request changes on any approval step of a report which is PENDING APPROVAL.

- After an engagement report has been approved by an advisor organization's approval chain, it is no longer automatically published to ANET: it gets the status APPROVED and is further visible only to it's author or to admins.
- As long as nothing happens the report stays in this state for a given period of time - we call it a quarantine period (which is configurable via anet.yml).
- During this period ANET administrators can return the report to the report author with comments (request changes) or decide to publish it earlier.

- The ANET administrators get an extra tab on their dashboard with all APPROVED reports (which means approved, but not yet published).
- The APPROVED reports on the dashboard of an ANET administrator are sorted ascending op updateDate. We actually want them to be sorted on approval timestamp, but directly after the approval action is added the report is updated too, so sorting on updated_at should result in the same order.

- When reading an APPROVED report one (the author of the report or the admin) can see when it is supposed to be published. The admin can also see a publish button which allows him/her to publish it before the quarantine period has passed and the report is being automatically published). The admin also has the possibility to request changes on an APPROVED report.
- When the admin has requested changes on an APPROVED report, the report has to go through the report approval chain again.

- APPROVED reports are retrieved as search results only if the user performing the search is their author or an ANET administrator.

- We now also display a subitem with approved reports under My Reports.
- Under My Reports we have now also added an item with Published reports and one with Cancelled reports (before they used to be one item, called published reports).
- Note that when we refer to the release of a report from now on we refer to the publication or the cancellation. The releasedAt property of a report and other related variables is relating to the publication or cancellation date.

- The e-mail notification about a report having been released to ANET is no longer sent when the report has been approved by an advisor organization's approval chain, but when the report has been published (by an ANET administrator or automatically by the system).

- We persist now each action on a report: submit, approval on an approval step, request changes on an approval step, request changes on an APPROVED report, publish action. The whole workflow of report actions are now being displayed for reports which are not in draft or future state. We should think better wether this is always wanted and the way the actions which are not related to an approval step are displayed should be improved.

### CODE REVIEW

File: src/main/resources/migrations.xml [hunk 1]
```diff
@@ -2234,4 +2234,40 @@
...
+			newTableName="reportActions"
+			oldTableName="approvalActions" />
+		<rollback>
+			<renameTable
+				newTableName="approvalActions"
+				oldTableName="reportActions" />
+		</rollback>
+	</changeSet>
+
+	<changeSet id="migrate_submit_and_publish_action" author="maradragan" dbms="mssql">
```
Comment (gjvoosten) [inline src/main/resources/migrations.xml]: Shouldn't there also be a migration to change the state of all previously RELEASED reports (`state = 2`) to APPROVED (`state = 6`)?

File: src/main/java/mil/dds/anet/database/ReportDao.java [hunk 2]
```diff
@@ -727,4 +737,44 @@ private Instant getRollupEngagmentStart(Instant start) {
...
+		ReportAction action = new ReportAction();
+		action.setReportUuid(r.getUuid());
+		if (user != null) {
+			//User is null when the publication action is being done automatically by a worker
+			action.setPersonUuid(user.getUuid());
+		}
+		action.setType(ActionType.PUBLISH);
+		AnetObjectEngine.getInstance().getReportActionDao().insert(action);
+
+		//Move the report to RELEASED state
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/database/ReportDao.java]: Update comment: `PUBLISHED state`

File: src/main/java/mil/dds/anet/database/ReportDao.java [hunk 3]
```diff
@@ -727,4 +737,44 @@ private Instant getRollupEngagmentStart(Instant start) {
...
+		r.setState(ReportState.PUBLISHED);
+		r.setReleasedAt(Instant.now());
+		final int numRows = this.update(r, r.getAuthor());
+		if (numRows != 0) {
+			sendReportPublisheddEmail(r);
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/database/ReportDao.java]: Rename to `sendReportPublishedEmail`

File: src/main/java/mil/dds/anet/database/ReportDao.java [hunk 4]
```diff
@@ -727,4 +737,44 @@ private Instant getRollupEngagmentStart(Instant start) {
 	public List<List<Task>> getTasks(List<String> foreignKeys) {
 		return tasksBatcher.getByForeignKeys(foreignKeys);
 	}
+
+	private void sendReportPublisheddEmail(Report r) {
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/database/ReportDao.java]: Rename to `sendReportPublishedEmail`

File: src/main/java/mil/dds/anet/search/sqlite/SqliteReportSearcher.java [hunk 5]
```diff
@@ -222,18 +223,28 @@ public SqliteReportSearcher() {
...
+				whereClauses.add("reports.state != :approvedState");
+				args.put("draftState", DaoUtils.getEnumId(ReportState.DRAFT));
+				args.put("rejectedState", DaoUtils.getEnumId(ReportState.REJECTED));
+				args.put("approvedState", DaoUtils.getEnumId(ReportState.APPROVED));
+			} else {
+				whereClauses.add("((reports.state != :draftState AND reports.state != :rejectedState) OR (reports.\"authorUuid\" = :userUuid))");
+				args.put("draftState", DaoUtils.getEnumId(ReportState.DRAFT));
+				args.put("rejectedState", DaoUtils.getEnumId(ReportState.REJECTED));
+				args.put("userUuid", user.getUuid());
+				if (AuthUtils.isAdmin(user) == false) {
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/search/sqlite/SqliteReportSearcher.java]: Change to `if (!AuthUtils.isAdmin(user)) {`

File: src/main/java/mil/dds/anet/threads/ReportPublicationWorker.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+	private void runInternal() {
+		//Get a list of all APPROVED reports
+		final ReportSearchQuery query = new ReportSearchQuery();
+		query.setPageSize(Integer.MAX_VALUE);
+		query.setState(Collections.singletonList(ReportState.APPROVED));
+		final List<Report> reports = dao.search(query, null, true).getList();
+		for (final Report r : reports) {
+			final List<ReportAction> workflow = r.loadWorkflow(context).join();
+			if (workflow.get(workflow.size()-1).getCreatedAt().isBefore(Instant.now().atZone(DaoUtils.getDefaultZoneId()).minusHours(this.nbOfHoursQuarantineApproved).toInstant())) {
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/threads/ReportPublicationWorker.java]: Extract to a local variable so the time you compare against is consistent throughout one internal run:
```
diff --git a/src/main/java/mil/dds/anet/threads/ReportPublicationWorker.java b/src/main/java/mil/dds/anet/threads/ReportPublicationWorker.java
index d21d297..b1aea9e 100644
--- a/src/main/java/mil/dds/anet/threads/ReportPublicationWorker.java
+++ b/src/main/java/mil/dds/anet/threads/ReportPublicationWorker.java
@@ -48,6 +48,7 @@ public class ReportPublicationWorker implements Runnable {
        }
 
        private void runInternal() {
+               final Instant now = Instant.now().atZone(DaoUtils.getDefaultZoneId()).minusHours(this.nbOfHoursQuarantineApproved).toInstant();
                //Get a list of all APPROVED reports
                final ReportSearchQuery query = new ReportSearchQuery();
                query.setPageSize(Integer.MAX_VALUE);
@@ -55,7 +56,7 @@ public class ReportPublicationWorker implements Runnable {
                final List<Report> reports = dao.search(query, null, true).getList();
                for (final Report r : reports) {
                        final List<ReportAction> workflow = r.loadWorkflow(context).join();
-                       if (workflow.get(workflow.size()-1).getCreatedAt().isBefore(Instant.now().atZone(DaoUtils.getDefaultZoneId()).minusHours(this.nbOfHoursQuarantineApproved).toInstant())) {
+                       if (workflow.get(workflow.size()-1).getCreatedAt().isBefore(now)) {
                                //Publish the report
                                try { 
                                        final int numRows = dao.publish(r, null);
```

Comment (gjvoosten) [inline src/main/java/mil/dds/anet/threads/ReportPublicationWorker.java]: What happens if `workflow.size()` is zero?

File: src/main/java/mil/dds/anet/threads/ReportPublicationWorker.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+				//Publish the report
+				try { 
+					final int numRows = dao.publish(r, null);
+					if (numRows == 0) {
+						throw new WebApplicationException("Couldn't process report publication", Status.NOT_FOUND);
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/threads/ReportPublicationWorker.java]: Don't throw a `WebApplicationException` here (it is not a user request), just log an error.

File: src/main/java/mil/dds/anet/threads/ReportPublicationWorker.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+		final ReportSearchQuery query = new ReportSearchQuery();
+		query.setPageSize(Integer.MAX_VALUE);
+		query.setState(Collections.singletonList(ReportState.APPROVED));
+		final List<Report> reports = dao.search(query, null, true).getList();
+		for (final Report r : reports) {
+			final List<ReportAction> workflow = r.loadWorkflow(context).join();
+			if (workflow.get(workflow.size()-1).getCreatedAt().isBefore(Instant.now().atZone(DaoUtils.getDefaultZoneId()).minusHours(this.nbOfHoursQuarantineApproved).toInstant())) {
+				//Publish the report
+				try { 
+					final int numRows = dao.publish(r, null);
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/threads/ReportPublicationWorker.java]: Shouldn't you be running this inside a transaction (as the `publish` method states)? So something like:
```
					final Handle dbHandle = AnetObjectEngine.getInstance().getDbHandle();
					dbHandle.inTransaction(h -> {
						final int numRows = dao.publish(r, null);
						if (numRows == 0) {
							logger.error("Couldn't process report publication");
						} else {
							AnetAuditLogger.log("report {} automatically published by the ReportPublicationWorker", r.getUuid());
						}
						return numRows;
					});
```

File: src/main/java/mil/dds/anet/search/sqlite/SqliteReportSearcher.java [hunk 9]
```diff
@@ -50,7 +51,7 @@ public SqliteReportSearcher() {
 		this("strftime('%%w', substr(reports.\"%s\", 1, 10)) + 1");	// %w day of week 0-6 with Sunday==0
 	}
 
-	public AnetBeanList<Report> runSearch(ReportSearchQuery query, Handle dbHandle, Person user) {
+	public AnetBeanList<Report> runSearch(ReportSearchQuery query, Handle dbHandle, Person user, Boolean systemSearch) {
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/search/sqlite/SqliteReportSearcher.java]: Change `Boolean` to a `boolean` primitive.

File: src/main/java/mil/dds/anet/search/mssql/MssqlReportSearcher.java [hunk 10]
```diff
@@ -33,12 +33,13 @@
...
 import mil.dds.anet.search.ReportSearchBuilder;
 import mil.dds.anet.search.ReportSearchBuilder.Comparison;
+import mil.dds.anet.utils.AuthUtils;
 import mil.dds.anet.utils.DaoUtils;
 import mil.dds.anet.utils.Utils;
 
 public class MssqlReportSearcher implements IReportSearcher {
 
-	public AnetBeanList<Report> runSearch(ReportSearchQuery query, Handle dbHandle, Person user) {
+	public AnetBeanList<Report> runSearch(ReportSearchQuery query, Handle dbHandle, Person user, Boolean systemSearch) {
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/search/mssql/MssqlReportSearcher.java]: Change `Boolean` to a `boolean` primitive.

File: src/main/java/mil/dds/anet/beans/Report.java [hunk 11]
```diff
@@ -455,61 +456,44 @@ public void setComments(List<Comment> comments) {
 						steps = getDefaultWorkflow(context, engine, defaultOrgUuid).join();
 					}
 				}
-				approvalStatus = createWorkflow(actions, steps);
+				List<ReportAction> newApprovalStepsActions = createApprovalStepsActions(actions, steps);
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/beans/Report.java]: Make `final`

File: src/main/java/mil/dds/anet/AnetApplication.java [hunk 12]
```diff
@@ -204,9 +205,15 @@ public void run(AnetConfiguration configuration, Environment environment) {
 		ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
 		AnetEmailWorker emailWorker = new AnetEmailWorker(engine.getEmailDao(), configuration, scheduler);
 		FutureEngagementWorker futureWorker = new FutureEngagementWorker(engine.getReportDao());
+		ReportPublicationWorker reportPublicationWorker = new ReportPublicationWorker(engine.getReportDao(), configuration);
 		
+		//Check for any reports that need to be published every 1 minutes.
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/AnetApplication.java]: Is it necessary to do this every minute (mind the plural 's' at the end)?

File: client/src/components/ReportWorkflow.js [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+const ActionDetails = ({action}) => {
+  if (action.type) {
+    return(
+      <div>
+        <span>By <LinkTo person={action.person} /></span><br/>
+        <small>
+          On {moment(action.createdAt).format(Settings.dateFormats.forms.short)}<br/>
+          At {moment(action.createdAt).format('h:mm a')}
```
Comment (gjvoosten) [inline client/src/components/ReportWorkflow.js]: Please use a date/time format from the dictionary; it was:
> On {moment(action.createdAt).format(Settings.dateFormats.forms.withTime)}
in `ReportApprovals.js`

File: client/src/pages/Home.js [hunk 14]
```diff
@@ -96,6 +97,11 @@ class BaseHome extends Page {
 		}
 	}
 
+	allApproved() { return {
```
Comment (gjvoosten) [inline client/src/pages/Home.js]: Move the `return` to the next line.

File: client/src/pages/Home.js [hunk 15]
```diff
@@ -147,42 +153,32 @@ class BaseHome extends Page {
...
-			tileFive: reportList(query: $queryFive) { totalCount },
-			savedSearches: mySearches {uuid, name, objectType, query}`
-		queries.forEach(q => {
+		let queryParts = []  // GQL query parts
+		queries.forEach((q, index) => {
 			q.query.pageNum = 0
 			q.query.pageSize = 1  // we're only interested in the totalCount, so just get at most one report
+			queryParts.push(
+				new GQL.Part(/* GraphQL */`tile`+index+`
+					: reportList(query:$query` + index + `) { totalCount}`)
```
Comment (gjvoosten) [inline client/src/pages/Home.js]: Simplify to:
``` new GQL.Part(/* GraphQL */`tile${index}: reportList(query:$query${index}) { totalCount}`) ```

File: src/main/java/mil/dds/anet/AnetApplication.java [hunk 16]
```diff
@@ -204,9 +205,15 @@ public void run(AnetConfiguration configuration, Environment environment) {
...
+		//And run once in 5 seconds from boot-up. (give the server time to boot up).
+		scheduler.scheduleAtFixedRate(reportPublicationWorker, 0, 1, TimeUnit.MINUTES);
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/AnetApplication.java]: Shouldn't the initial delay be larger than `0`?

File: src/main/java/mil/dds/anet/threads/ReportPublicationWorker.java (hunk 6 above)
Comment (maradragan) [reply src/main/java/mil/dds/anet/threads/ReportPublicationWorker.java]: I thought, if the report is approved it must have a workflow... ?

Comment (gjvoosten) [reply src/main/java/mil/dds/anet/threads/ReportPublicationWorker.java]: Better check (and possibly log an error) than assume.

File: src/main/java/mil/dds/anet/AnetApplication.java (hunk 16 above)
Comment (maradragan) [reply src/main/java/mil/dds/anet/AnetApplication.java]: I will make it 5.

**Pull request merged**